### PR TITLE
:tada: engineering: add prefect engine when running ETL

### PR DESCRIPTION
**Pull request by Marigold**
Adds option `--engine` for specifying which scheduler to use (default is `--engine etl`). Using `--engine prefect` orchestrates ETL with Prefect and creates SQLite file that could be inspected with Prefect UI.

The Prefect UI runs on http://staging-site-prefect:4200/flow-runs (there's a new link from Wizard). Here's an [example of a run](http://staging-site-prefect:4200/flow-runs/flow-run/88d05e88-0371-482e-bb3c-82c8e1abd950) after changing regions.

It runs steps **concurrently** with a single worker (default) and uses Dask with multiple workers (flag `--workers`).

## Comparison to `--engine etl`
- Prefect doesn't interrupt other tasks on error and tries to complete as many tasks as possible
- When using multiple workers, it's easy to find the failing task and the exception (it's not as trivial from Buildkite logs)
- `structlog.info` adds colour to output, but Prefect can't decode it and prints characters like `[0m [[32m[1minfo`. This should be soon fixed in https://github.com/PrefectHQ/prefect-ui-library/pull/2582
- There's not much overhead from Prefect compared to multiprocessing
- We could leverage Dask features (e.g. memory limiting), but it's also likely that a naive multiprocessing is good enough for us and we shouldn't complicate our lives (like with DVC)

## My 2 cents

I'd find it very helpful for inspecting ETL runs on staging servers and in production (where I find searching through logs really annoying). We could give it a try and see if it was worth it in a few weeks. Prefect could also be useful for automatic dataset updates, which currently exist as bash scripts and are run by Buildkite. It works, but as @lucasrodes suggested, we might need more flexibility.

## TODO before merging
- [ ] Undo changes to regions.yml

### CODE REVIEW

**Comment by lucasrodes** [comment]
This looks great, thanks for doing this, Mojmir.

This would be an alternative to Buildkite, where error logs are more legible?

I tried checking the Prefect UI on Wizard, and realised that one needs to first run `make prefect-ui`. This feels a bit confusing at first, maybe we should signal this somewhere. Maybe Wizard can have an option to trigger it? or some message Also wondering if the app should be in the 'Misc' category, or rather in 'Monitoring', for instance.

Also, I tried running `make prefect-ui`, but run into the following error.


```
make prefect-ui
--- Starting Prefect UI at http://0.0.0.0:4200
poetry run prefect server start --host 0.0.0.0

 ___ ___ ___ ___ ___ ___ _____ 
| _ \ _ \ __| __| __/ __|_   _| 
|  _/   / _|| _|| _| (__  | |  
|_| |_|_\___|_| |___\___| |_|  

Configure Prefect to communicate with the server with:

    prefect config set PREFECT_API_URL=http://0.0.0.0:4200/api

View the API reference documentation at http://0.0.0.0:4200/docs

Check out the dashboard at http://0.0.0.0:4200



10:17:43.352 | ERROR   | prefect.server.services.flowrunnotifications - Unexpected error in: OperationalError('(sqlite3.OperationalError) database is locked')
Traceback (most recent call last):
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 146, in execute
    self._adapt_connection._handle_exception(error)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 298, in _handle_exception
    raise error
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 128, in execute
    self.await_(_cursor.execute(operation, parameters))
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 131, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 48, in execute
    await self._execute(self._cursor.execute, sql, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 40, in _execute
    return await self._conn._execute(fn, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 132, in _execute
    return await future
           ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 115, in run
    result = function()
             ^^^^^^^^^^
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/services/loop_service.py", line 79, in start
    await self.run_once()
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/dependencies.py", line 125, in async_wrapper
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py", line 222, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py", line 222, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/configurations.py", line 453, in begin_transaction
    yield transaction
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/interface.py", line 119, in session_context
    yield session
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/services/flow_run_notifications.py", line 38, in run_once
    notifications = await db.get_flow_run_notifications_from_queue(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/interface.py", line 405, in get_flow_run_notifications_from_queue
    return await self.queries.get_flow_run_notifications_from_queue(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/query_components.py", line 1115, in get_flow_run_notifications_from_queue
    await session.execute(delete_stmt)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/session.py", line 461, in execute
    result = await greenlet_spawn(
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 201, in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2351, in execute
    return self._execute_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2236, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/bulk_persistence.py", line 1953, in orm_execute_statement
    return super().orm_execute_statement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/context.py", line 293, in orm_execute_statement
    result = conn.execute(
             ^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 146, in execute
    self._adapt_connection._handle_exception(error)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 298, in _handle_exception
    raise error
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 128, in execute
    self.await_(_cursor.execute(operation, parameters))
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 131, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 48, in execute
    await self._execute(self._cursor.execute, sql, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 40, in _execute
    return await self._conn._execute(fn, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 132, in _execute
    return await future
           ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 115, in run
    result = function()
             ^^^^^^^^^^
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: DELETE FROM flow_run_notification_queue WHERE flow_run_notification_queue.id IN (SELECT 1 FROM (SELECT 1) WHERE 1!=1) RETURNING id]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

**Comment by larsyencken** [comment]
> This would be an alternative to Buildkite, where error logs are more legible?

This is more of a complement. We would run the Prefect web UI every staging server for you, and it would be an additional way of seeing individual ETL runs or changes made to your staging server and their logs.

**Comment by Marigold** [comment]
> Also, I tried running make prefect-ui, but run into the following error.

Hm, weird. Could you try running `make prefect-reset` and if that doesn't work, remove file `~/.prefect/prefect.db`?

> Also wondering if the app should be in the 'Misc' category, or rather in 'Monitoring', for instance.

Oh right 🤦 Moved it there.

> This feels a bit confusing at first, maybe we should signal this somewhere. Maybe Wizard can have an option to trigger it? or some message

Good idea! Fixed that with nicer error message.

> Some entry in the docs could be of great use if we want data managers / engineers to use this tool.

The plan is to "dark launch" this first. Once we confirm that it has value, we should let data managers know and add it to docs.

**Comment by larsyencken** [comment]
@Marigold Marcel and I were trying this now, and overall it's quite impressive, though there's a ton of errors of this form littered through the run, preventing it from executing cleanly:

```
Encountered exception during execution:
Traceback (most recent call last):
  File "/Users/lars/Documents/owid/etl/.venv/lib/python3.11/site-packages/prefect/engine.py", line 2145, in orchestrate_task_run
    result = await call.aresult()
             ^^^^^^^^^^^^^^^^^^^^
  File "/Users/lars/Documents/owid/etl/.venv/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 327, in aresult
    return await asyncio.wrap_future(self.future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lars/Documents/owid/etl/.venv/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 352, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lars/Documents/owid/etl/etl/command.py", line 421, in run_step
    return step.run()
           ^^^^^^^^^^
  File "/Users/lars/Documents/owid/etl/etl/steps/__init__.py", line 446, in run
    dataset.save()
  File "/Users/lars/Documents/owid/etl/lib/catalog/owid/catalog/datasets.py", line 173, in save
    assert self.metadata.short_name, "Missing dataset short_name"
AssertionError: Missing dataset short_name
```

**Comment by larsyencken** [comment]
💡 If we got this working and integrated this, one cool thing is that we could potentially provide step-level and run-level analytics on all our ETL steps and production runs, basically telling you how things are changing over time and what's most expensive.

File: etl/command.py [hunk 1]
```diff
@@ -357,16 +376,53 @@ def run_dag(
...
+                print("Failure detected. Shutting down Dask client and cancelling all tasks.")
+                # force kill Dask client
+                client.shutdown()
+
+        @prefect.flow(log_prints=True, task_runner=task_runner)
+        def run_etl(steps):
+            task_futures: Dict[str, PrefectFuture] = {}
+
+            for step in steps:
+                # task = prefect.task(name=str(step), on_failure=[on_failure_hook])
```
Comment (larsyencken) [inline etl/command.py:411]: I am noticing that the flows hang around forever on failure. Is there a reason this bit is commented out?

File: etl/command.py [hunk 2]
```diff
@@ -357,16 +376,53 @@ def run_dag(
...
+                # include dependencies that are in the list of steps
+                wait_for = [task_futures[str(dep)] for dep in step.dependencies if str(dep) in task_futures]
+
+                def run_step(grapher_insert_workers=config.GRAPHER_INSERT_WORKERS):
+                    # We dynamically update `GRAPHER_INSERT_WORKERS` based on number of processes, the change in config
+                    # is not propagated when Dask creates a new process. We have to set it inside each task.
+                    config.GRAPHER_INSERT_WORKERS = grapher_insert_workers
+                    return step.run()
+
+                task_futures[str(step)] = task(run_step).submit(wait_for=wait_for)  # type: ignore
```
Comment (larsyencken) [inline etl/command.py:423]: You should have a separate run method for each engine, rather than a massive block like this.

**Review by larsyencken — CHANGES_REQUESTED** [review]
Overall it looks super nice, mainly the many task failures when executing it would need to be solved.

**Comment by Marigold** [comment]
> though there's a ton of errors of this form littered through the run, preventing it from executing cleanly:

@larsyencken how did you get the error? I can't replicate it...

**Comment by lucasrodes** [comment]
@Marigold Thanks for the suggestions.

Removing `~/.prefect/prefect.db` made it work!


Tried `make prefect-reset`, but got a slightly different error now:

```
make prefect-reset
==> Installing packages
poetry install --no-ansi || poetry install --no-ansi
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: etl (0.1.0)
touch .venv
--- Resetting Prefect database
poetry run prefect server database reset -y
Downgrading database...
Upgrading database...
/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py:144: SAWarning: Skipped unsupported reflection of expression-based index ix_flow_run__coalesce_start_time_expected_start_time_desc
  next(self.gen)
/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py:144: SAWarning: Skipped unsupported reflection of expression-based index ix_flow_run__coalesce_start_time_expected_start_time_asc
  next(self.gen)
Prefect database "sqlite+aiosqlite:////home/lucas/.prefect/prefect.db" reset!
etl-py3.11(etl-py3.11) ➜  etl git:(prefect) ✗ make prefect-ui
--- Starting Prefect UI at http://0.0.0.0:4200
poetry run prefect server start --host 0.0.0.0

 ___ ___ ___ ___ ___ ___ _____ 
| _ \ _ \ __| __| __/ __|_   _| 
|  _/   / _|| _|| _| (__  | |  
|_| |_|_\___|_| |___\___| |_|  

Configure Prefect to communicate with the server with:

    prefect config set PREFECT_API_URL=http://0.0.0.0:4200/api

View the API reference documentation at http://0.0.0.0:4200/docs

Check out the dashboard at http://0.0.0.0:4200



13:48:05.689 | ERROR   | prefect.server.services.telemetry - Unexpected error in: OperationalError('(sqlite3.OperationalError) database is locked')
Traceback (most recent call last):
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 146, in execute
    self._adapt_connection._handle_exception(error)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 298, in _handle_exception
    raise error
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 128, in execute
    self.await_(_cursor.execute(operation, parameters))
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 131, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 48, in execute
    await self._execute(self._cursor.execute, sql, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 40, in _execute
    return await self._conn._execute(fn, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 132, in _execute
    return await future
           ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/core.py", line 115, in run
    result = function()
             ^^^^^^^^^^
sqlite3.OperationalError: database is locked

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/services/loop_service.py", line 79, in start
    await self.run_once()
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/services/telemetry.py", line 85, in run_once
    await self._fetch_or_set_telemetry_session()
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/dependencies.py", line 125, in async_wrapper
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py", line 222, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/home/lucas/.pyenv/versions/3.11.1/lib/python3.11/contextlib.py", line 222, in __aexit__
    await self.gen.athrow(typ, value, traceback)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/configurations.py", line 453, in begin_transaction
    yield transaction
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/interface.py", line 119, in session_context
    yield session
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/services/telemetry.py", line 63, in _fetch_or_set_telemetry_session
    await configuration.write_configuration(session, telemetry_session)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/database/dependencies.py", line 125, in async_wrapper
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/prefect/server/models/configuration.py", line 29, in write_configuration
    await session.flush()
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/ext/asyncio/session.py", line 800, in flush
    await greenlet_spawn(self.sync_session.flush, objects=objects)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 203, in greenlet_spawn
    result = context.switch(value)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4341, in flush
    self._flush(objects)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4476, in _flush
    with util.safe_reraise():
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4437, in _flush
    flush_context.execute()
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 93, in save_obj
    _emit_insert_statements(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 1233, in _emit_insert_statements
    result = connection.execute(
             ^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 146, in execute
    self._adapt_connection._handle_exception(error)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 298, in _handle_exception
    raise error
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/dialects/sqlite/aiosqlite.py", line 128, in execute
    self.await_(_cursor.execute(operation, parameters))
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 131, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site-packages/aiosqlite/cursor.py", line 48, in execute
    await self._execute(self._cursor.execute, sql, parameters)
  File "/home/lucas/repos/etl/.venv/lib/python3.11/site
```

**Comment by Marigold** [comment]
@lucasrodes is the dashboard running on http://0.0.0.0:4200? I read that the timeout thing is more of a warning than an actual error.

File: etl/command.py (hunk 1 above)
Comment (Marigold) [reply etl/command.py]: Hmm, when does it happen? When I run it locally with an error, it doesn't hang **forever**, but it fails after all tasks are either completed or failed.

**Comment by lucasrodes** [comment]
@Marigold after removing `~/.prefect/prefect.db`, it works perfectly. However, if I then run 

```
make prefect-reset
make prefect-ui
```

I get the `sqlalchemy.exc.OperationalError` error I mentioned [above](https://github.com/owid/etl/pull/3029#issuecomment-2259941396).

```
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) database is locked
[SQL: DELETE FROM flow_run_notification_queue WHERE flow_run_notification_queue.id IN (SELECT 1 FROM (SELECT 1) WHERE 1!=1) RETURNING id]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

**Comment by Marigold** [comment]
> though there's a ton of errors of this form littered through the run, preventing it from executing cleanly:

@larsyencken how did you get the error, please? I can't replicate it.

**Comment by Marigold** [comment]
This is ready to be merged, but I'm rethinking whether it's a good idea to introduce more complexity to ETL. Setting it as blocked until we find the right moment.

**Comment by lucasrodes** [comment]
@Marigold that makes sense. I've unset myself from the reviewers list. Feel free to add me back once you want me to review it! Thanks <3

**Comment by Marigold** [comment]
Even though I'm in love with prefect, there's not as much added value and likely not worth making our stack more complex.

**Pull request closed without merging**